File: pxr/usd/sdf/testenv/testSdfFileFormatCapabilities.py
#!/pxrpythonsubst
#
# Copyright 2023 Pixar
#
# Licensed under the terms set forth in the LICENSE.txt file available at
# https://openusd.org/license.

from __future__ import print_function
import os, tempfile, unittest
from pxr import Plug, Sdf, Tf

class TestSdfCapabilities(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        # Register dso plugins.
        testRoot = os.path.join(os.path.dirname(__file__), 'SdfPlugins')
        testPluginsDso = testRoot + '/lib'
        testPluginsDsoSearch = testPluginsDso + '/*/Resources/'
        Plug.Registry().RegisterPlugins(testPluginsDsoSearch)

    # sanity check for existing file formats which specify nothing in the
    # plugInfo.json file.
    def test_StaticDefault(self):
        ext = '.sdf'
        self.assertTrue(Sdf.FileFormat.FormatSupportsReading(ext))
        self.assertTrue(Sdf.FileFormat.FormatSupportsWriting(ext))
        self.assertTrue(Sdf.FileFormat.FormatSupportsEditing(ext))

    def test_StaticUnwritable(self):
        ext = '.unwritable'
        target = 'sdf'
        self.assertTrue(Sdf.FileFormat.FormatSupportsReading(ext))
        self.assertFalse(Sdf.FileFormat.FormatSupportsWriting(ext))
        self.assertTrue(Sdf.FileFormat.FormatSupportsEditing(ext))

        self.assertTrue(Sdf.FileFormat.FormatSupportsReading(ext, target))
        self.assertFalse(Sdf.FileFormat.FormatSupportsWriting(ext, target))
        self.assertTrue(Sdf.FileFormat.FormatSupportsEditing(ext, target))

    def test_StaticUnreadable(self):
        ext = '.unreadable'
        target = 'test'
        self.assertFalse(Sdf.FileFormat.FormatSupportsReading(ext))
        self.assertTrue(Sdf.FileFormat.FormatSupportsWriting(ext))
        self.assertTrue(Sdf.FileFormat.FormatSupportsEditing(ext))

        self.assertFalse(Sdf.FileFormat.FormatSupportsReading(ext, target))
        self.assertTrue(Sdf.FileFormat.FormatSupportsWriting(ext, target))
        self.assertTrue(Sdf.FileFormat.FormatSupportsEditing(ext, target))

    def test_StaticInvalid(self):
        ext = '.bogus'
        self.assertFalse(Sdf.FileFormat.FormatSupportsReading(ext))
        self.assertFalse(Sdf.FileFormat.FormatSupportsWriting(ext))
        self.assertFalse(Sdf.FileFormat.FormatSupportsEditing(ext))

        ext = '.unreadable'
        target = 'invalid'
        self.assertFalse(Sdf.FileFormat.FormatSupportsReading(ext, target))
        self.assertFalse(Sdf.FileFormat.FormatSupportsWriting(ext, target))
        self.assertFalse(Sdf.FileFormat.FormatSupportsEditing(ext, target))

    def test_InstanceDefault(self):
        sdfFileFormat = Sdf.FileFormat.FindByExtension('.sdf')
        self.assertIsNotNone(sdfFileFormat)
        self.assertTrue(sdfFileFormat.SupportsReading())
        self.assertTrue(sdfFileFormat.SupportsWriting())
        self.assertTrue(sdfFileFormat.SupportsEditing())

    def test_InstanceUnwritable(self):
        unwritableFileFormat = Sdf.FileFormat.FindByExtension('.unwritable')
        self.assertIsNotNone(unwritableFileFormat)
        self.assertTrue(unwritableFileFormat.SupportsReading())
        self.assertFalse(unwritableFileFormat.SupportsWriting())
        self.assertTrue(unwritableFileFormat.SupportsEditing())

    def test_InstanceUnreadable(self):
        unreadableFileFormat = Sdf.FileFormat.FindByExtension('.unreadable')
        self.assertIsNotNone(unreadableFileFormat)
        self.assertFalse(unreadableFileFormat.SupportsReading())
        self.assertTrue(unreadableFileFormat.SupportsWriting())
        self.assertTrue(unreadableFileFormat.SupportsEditing())

    def test_InstanceNonEditable(self):
        uneditableFileFormat = Sdf.FileFormat.FindByExtension('.uneditable')
        self.assertIsNotNone(uneditableFileFormat)

        self.assertTrue(uneditableFileFormat.SupportsReading())
        self.assertTrue(uneditableFileFormat.SupportsWriting())
        self.assertFalse(uneditableFileFormat.SupportsEditing())

    def test_export(self):
        layer = Sdf.Layer.CreateAnonymous()
        with self.assertRaises(Tf.ErrorException):
            layer.Export('file.unwritable')

    def test_import(self):
        # use of tempfile here ensures valid filesystem path
        with tempfile.NamedTemporaryFile(suffix='.writeonly') as temp:
            layer = Sdf.Layer.CreateAnonymous()

            with self.assertRaises(Tf.ErrorException):
                    layer.Import(temp.name)


if __name__ == "__main__":
    unittest.main()

File: pxr/usd/sdf/testenv/TestSdfFileFormatCapabilities.cpp
//
// Copyright 2023 Pixar
//
// Licensed under the terms set forth in the LICENSE.txt file available at
// https://openusd.org/license.
//
#include "pxr/pxr.h"
#include "pxr/base/tf/staticTokens.h"
#include "pxr/usd/sdf/layer.h"
#include "pxr/usd/sdf/usdaFileFormat.h"

PXR_NAMESPACE_OPEN_SCOPE

#define TEST_SDF_UNWRITABLE_FILE_FORMAT_TOKENS       \
    ((Extension, "unwritable")) \
    ((VersionStr, "0.0.0")) \
    ((Target, "")) 

TF_DECLARE_PUBLIC_TOKENS(TestSdfUnwritableFormat_Tokens, 
                         TEST_SDF_UNWRITABLE_FILE_FORMAT_TOKENS);

TF_DEFINE_PUBLIC_TOKENS(
    TestSdfUnwritableFormat_Tokens, 
    TEST_SDF_UNWRITABLE_FILE_FORMAT_TOKENS);

class TestSdfUnwritableFormat : public SdfUsdaFileFormat
{
public:
    SDF_FILE_FORMAT_FACTORY_ACCESS;

    TestSdfUnwritableFormat()
        :SdfUsdaFileFormat(
            TestSdfUnwritableFormat_Tokens->Extension,
            TestSdfUnwritableFormat_Tokens->VersionStr,
            TestSdfUnwritableFormat_Tokens->Target)
    {
    }
};

TF_REGISTRY_FUNCTION(TfType)
{
    SDF_DEFINE_FILE_FORMAT(TestSdfUnwritableFormat, 
                           SdfUsdaFileFormat);
}

#define TEST_SDF_UNREADABLE_FILE_FORMAT_TOKENS       \
    ((Extension, "unreadable")) \
    ((VersionStr, "0.0.0")) \
    ((Target, "test")) 

TF_DECLARE_PUBLIC_TOKENS(TestSdfUnreadableFormat_Tokens, 
                         TEST_SDF_UNREADABLE_FILE_FORMAT_TOKENS);

TF_DEFINE_PUBLIC_TOKENS(
    TestSdfUnreadableFormat_Tokens, 
    TEST_SDF_UNREADABLE_FILE_FORMAT_TOKENS);

class TestSdfUnreadableFormat : public SdfUsdaFileFormat
{
public:
    SDF_FILE_FORMAT_FACTORY_ACCESS;

    TestSdfUnreadableFormat()
        :SdfUsdaFileFormat(
            TestSdfUnreadableFormat_Tokens->Extension,
            TestSdfUnreadableFormat_Tokens->VersionStr,
            TestSdfUnreadableFormat_Tokens->Target)
    {
    }
};

TF_REGISTRY_FUNCTION(TfType)
{
    SDF_DEFINE_FILE_FORMAT(TestSdfUnreadableFormat, 
                           SdfUsdaFileFormat);
}


#define TEST_SDF_UNEDITABLE_FILE_FORMAT_TOKENS       \
    ((Extension, "uneditable")) \
    ((VersionStr, "0.0.0")) \
    ((Target, "test")) 

TF_DECLARE_PUBLIC_TOKENS(TestSdfUneditableFormat_Tokens, 
                         TEST_SDF_UNEDITABLE_FILE_FORMAT_TOKENS);

TF_DEFINE_PUBLIC_TOKENS(
    TestSdfUneditableFormat_Tokens, 
    TEST_SDF_UNEDITABLE_FILE_FORMAT_TOKENS);

class TestSdfUneditableFormat : public SdfUsdaFileFormat
{
public:
    SDF_FILE_FORMAT_FACTORY_ACCESS;

    TestSdfUneditableFormat()
        :SdfUsdaFileFormat(
            TestSdfUneditableFormat_Tokens->Extension,
            TestSdfUneditableFormat_Tokens->VersionStr,
            TestSdfUneditableFormat_Tokens->Target)
    {
    }
};

TF_REGISTRY_FUNCTION(TfType)
{
    SDF_DEFINE_FILE_FORMAT(TestSdfUneditableFormat, 
                           SdfUsdaFileFormat);
}


PXR_NAMESPACE_CLOSE_SCOPE



